src/Meshes/Meshes.jl

```diff
@@ -108,4 +111,113 @@ struct EquiangularCubedSphereMesh{FT} <: AbstractMesh{FT}
     n::Int64
 end
 
+"""
+    TensorProductMesh(domain::RectangleDomain, n::Integer)
+
+A tensor-product `AbstractMesh` of `domain` with `n1` elements in dimension 1, and `n2`
+in dimension 2.
+"""
+struct TensorProductMesh{FT, RD <: RectangleDomain{FT}} <: AbstractMesh{FT}
+    domain::RD
+    n1::Int64 # number of elements in x1 direction
+    n2::Int64 # number of elements in x2 direction
+    faces::Vector{Tuple{Int64, Int64, Int64, Int64, Bool}}
+    coordinates::Vector{Cartesian2DPoint{FT}}
+end
+
+function TensorProductMesh(
+    domain::RectangleDomain{FT},
+    n1,
+    n2,
+    coordinates = nothing,
+) where {FT}
+
+    nelem = n1 * n2
+    x1periodic = isnothing(domain.x1boundary)
+    x2periodic = isnothing(domain.x2boundary)
+    faces = Vector{Tuple{Int64, Int64, Int64, Int64, Bool}}(undef, nelem * 4)
+
+    # Store all mesh faces as (elem, face, opelem, opface, reversed)
+    # so that we can determine face pairs via the map (elem, face) to neighbouring (opelem, opface, reversed)
+    for e in 1:nelem
+        z2s, z1s = fldmod(e - 1, n1)
+
+        # Face 1
+        z1 = z1s - 1
+        z2 = z2s
+        if z1 < 0 && !x1periodic
+            faces[(e - 1) * 4 + 1] = (e, 1, 0, 1, false)
+        else
+            if z1 < 0
+                z1 += n1
+            end
+            opface = 2
+            opelem = z2 * n1 + z1 + 1
+            faces[(e - 1) * 4 + 1] = (e, 1, opelem, opface, false)
+        end
+
+        # Face 2
+        z1 = z1s + 1
+        if z1 == n1 && !x1periodic
+            faces[(e - 1) * 4 + 2] = (e, 2, 0, 2, false)
+        else
+            if z1 == n1
+                z1 -= n1
+            end
+            opface = 1
+            opelem = z2 * n1 + z1 + 1
+            faces[(e - 1) * 4 + 2] = (e, 2, opelem, opface, false)
+        end
+
+        # Face 3
+        z1 = z1s
+        z2 = z2s - 1
+        if z2 < 0 && !x2periodic
+            faces[(e - 1) * 4 + 3] = (e, 3, 0, 3, false)
+        else
+            if z2 < 0
+                z2 += n2
+            end
+            opface = 4
+            opelem = z2 * n1 + z1 + 1
+            faces[(e - 1) * 4 + 3] = (e, 3, opelem, opface, false)
+        end
+
+        # Face 4
+        z2 = z2s + 1
+        if z2 == n2 && !x2periodic
+            faces[(e - 1) * 4 + 4] = (e, 4, 0, 4, false)
+        else
+            if z2 == n2
+                z2 -= n2
+            end
+            opface = 3
+            opelem = z2 * n1 + z1 + 1
+            faces[(e - 1) * 4 + 4] = (e, 4, opelem, opface, false)
+        end
+    end
+
+    if isnothing(coordinates)
+
+        coordinates = Vector{Cartesian2DPoint{FT}}(undef, (n1 + 1) * (n2 + 1))
+        # Default equispaced vertex coordinates, if the user has not specified their locations
+        range1 = range(domain.x1min, domain.x1max; length = n1 + 1)
+        range2 = range(domain.x2min, domain.x2max; length = n2 + 1)
+
+        # Coordinates array, row-major storage
```

Comment:
why are we storing these in row major?  shouldn't we use the same ND `DataLayouts` storage backing as the existing coordinate representation (and what the rest of the array ds use)

Reply:
Probably yes, I just wasn't familiar enough to use the existing DataLayouts. This was probably going to change anyways, once we had defined how and where to define the warping function. So this was only a quick implementation to test the existing behavior. This block of code might change anyways in view of the latest notes in #75

Reply:
When I chatted with @simonbyrne he said that we can leave this as is for now, since it is more pressing to get to a cubed-sphere mesh, for which this PR is an intermediate step than optimizing code. Unless you have a strong preference about this -- in that case, I might need some walk-through the existing DataLayouts storage to get familiar with it :)